cpp/tests/prims/mg_transform_reduce_v.cu

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utilities/high_res_clock.h>
+#include <utilities/base_fixture.hpp>
+#include <utilities/device_comm_wrapper.hpp>
+#include <utilities/test_graphs.hpp>
+#include <utilities/test_utilities.hpp>
+#include <utilities/thrust_wrapper.hpp>
+
+#include <cugraph/algorithms.hpp>
+#include <cugraph/partition_manager.hpp>
+
+#include <cuco/detail/hash_functions.cuh>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/prims/transform_reduce_v.cuh>
+
+#include <thrust/count.h>
+#include <raft/comms/comms.hpp>
+#include <raft/comms/mpi_comms.hpp>
+#include <raft/handle.hpp>
+#include <rmm/device_scalar.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <gtest/gtest.h>
+
+#include <random>
+
+template <typename vertex_t, typename T>
+struct property_transform : public thrust::unary_function<vertex_t, T> {
+  int mod{};
+  property_transform(int mod_count) : mod(mod_count) {}
+  __device__ auto operator()(const vertex_t& val)
+  {
+    cuco::detail::MurmurHash3_32<vertex_t> hash_func{};
+    auto value = hash_func(val) % mod;
+    return static_cast<T>(value);
+  }
+};
+
+template <typename vertex_t, typename... T>
+struct property_transform<vertex_t, std::tuple<T...>>
```

Comment:
The issue with `thrust::tuple` is that is mangles the parameter pack. So for example, if I were to pass `thrust::tuple<int, float>` the parameter pack won't be `int, float`. It would be `int, float, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type`. When you try to use the parameter pack in the template, the compiler tries to instantiate thrust::null_type when you use fold expressions.

In this case, I am using `std::tuple` only to work with the parameter pack. The operator being instantiated though is still a unary function that returns a `thrust::tuple`.

template <typename vertex_t, typename... T>
struct property_transform<vertex_t, **std::tuple<T...>**>
  : public thrust::unary_function<vertex_t, **thrust::tuple<T...>**>

Reply:
Gotcha.